signac/core/synced_collection.py

```diff
@@ -0,0 +1,164 @@
+        self._data = None
+        self._parent = parent
+        self._suspend_sync_ = 0
+
+    @classmethod
+    def register(cls, *args):
+        """Register the synced data structures.
+
+        Registry is used when recursively converting synced data structures to determine
+        what to convert their children into.
+
+        Parameters
+        ----------
+        *args
+            Classes to register
+        """
+        if not hasattr(cls, 'registry'):
+            cls.registry = defaultdict(list)
+        for _cls in args:
+            cls.registry[_cls.backend].append(_cls)
+
+    @classmethod
+    def from_base(cls, data, backend=None, **kwargs):
+        """Dynamically resolve the type of object to the corresponding synced collection.
+
+        Parameters
+        ----------
+        data : any
+            Data to be converted from base class.
+        backend: str
+            Name of backend for synchronization. Default to backend of class.
+        **kwargs:
+            Kwargs passed to instance of synced collection.
+
+        Returns
+        -------
+        data : object
+            Synced object of corresponding base type.
+        """
+        backend = cls.backend if backend is None else backend
+        if backend is None:
+            raise ValueError("No backend found.")
+        for _cls in cls.registry[backend]:
+            if _cls.is_base_type(data):
+                return _cls(data=data, **kwargs)
+        if NUMPY:
+            if isinstance(data, numpy.number):
+                return data.item()
+        return data
+
+    @abstractmethod
+    def to_base(self):
+        """Dynamically resolve the synced collection to the corresponding base type."""
+        pass
+
+    @contextmanager
+    def _suspend_sync(self):
+        """Prepare context where load and sync are suspended."""
+        self._suspend_sync_ += 1
+        yield
+        self._suspend_sync_ -= 1
+
+    @classmethod
+    @abstractmethod
+    def is_base_type(cls, data):
+        """Check whether data is of the same base type (such as list or dict) as this class."""
+        pass
+
+    @abstractmethod
+    def _load(self):
+        """Load data from underlying backend."""
+        pass
+
+    @abstractmethod
+    def _sync(self):
+        """Write data to underlying backend."""
+        pass
+
+    def sync(self):
+        """Synchronize the data with the underlying backend."""
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                self._sync()
+            else:
+                self._parent.sync()
+
+    def load(self):
+        """Load the data from the underlying backend."""
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                data = self._load()
+                with self._suspend_sync():
+                    self._update(data)
+            else:
+                self._parent.load()
+
+    # methods having same implementaion for all data-structures
```

Comment:
Your comment is still not providing sufficient explanation.
```suggestion
    # The following methods share a common implementation for all data structures and regardless of backend.
```
Is this correct?

Reply:
Also, please put in more effort in reviewing your own diffs and commits to avoid the common occurrence of typos etc. (`implementaion`).